pkg/routing/config_map_updater.go

```diff
@@ -21,49 +19,40 @@ import (
 //	called ConfigMapRoutingTableName. On either of those events, decodes
 //	that ConfigMap into a routing table and stores the new table into table
 //	using table.Replace(newTable)
+//	- Execute the callback function, if one exists
 //	- Returns an appropriate non-nil error if ctx.Done() receives
 func StartConfigMapRoutingTableUpdater(
 	ctx context.Context,
 	lggr logr.Logger,
-	updateEvery time.Duration,
-	getterWatcher k8s.ConfigMapGetterWatcher,
+	cmInformer *k8s.InformerConfigMapUpdater,
+	ns string,
 	table *Table,
 	q queue.Counter,
+	cbFunc func() error,
 ) error {
 	lggr = lggr.WithName("pkg.routing.StartConfigMapRoutingTableUpdater")
-	watchIface, err := getterWatcher.Watch(ctx, metav1.ListOptions{})
-	if err != nil {
-		return err
-	}
-	defer watchIface.Stop()
 
-	ticker := time.NewTicker(updateEvery)
-	defer ticker.Stop()
-	for {
-		select {
-		case <-ctx.Done():
-			return errors.Wrap(ctx.Err(), "context is done")
-		case <-ticker.C:
-			if err := GetTable(ctx, lggr, getterWatcher, table, q); err != nil {
-				return errors.Wrap(err, "failed to fetch routing table")
-			}
+	watcher := cmInformer.Watch(ns, ConfigMapRoutingTableName)
+	defer watcher.Stop()
 
-		case evt := <-watchIface.ResultChan():
-			evtType := evt.Type
-			obj := evt.Object
-			if evtType == watch.Added || evtType == watch.Modified {
-				cm, ok := obj.(*corev1.ConfigMap)
-				// by definition of watchIface, all returned objects should
-				// be assertable to a ConfigMap. In the likely impossible
-				// case that it isn't, just ignore and move on.
-				// This check is just to be defensive.
+	grp, ctx := errgroup.WithContext(ctx)
+	defer ctx.Done()
```

Comment:
`defer ctx.Done()` is effectively a no-op here. I think a better way to go would be the below code. you'd also need to replace the other `defer ctx.Done()` lines with `defer done()`

```suggestion
        ctx, done := context.WithCancel(ctx)
        defer done()
	grp, ctx := errgroup.WithContext(ctx)
```

Reply:
Thanks! I've learned.